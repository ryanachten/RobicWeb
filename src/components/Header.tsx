import * as React from "react";
import { Link } from "react-router-dom";
import AppBar from "@material-ui/core/AppBar";
import Drawer from "@material-ui/core/Drawer";
import IconButton from "@material-ui/core/IconButton";
import List from "@material-ui/core/List";
import ListItem from "@material-ui/core/ListItem";
import ListItemText from "@material-ui/core/ListItemText";
import MenuIcon from "@material-ui/icons/Menu";
import Toolbar from "@material-ui/core/Toolbar";
import { Classes } from "jss";
import createStyles from "@material-ui/core/styles/createStyles";
import withStyles, { WithStyles } from "@material-ui/core/styles/withStyles";
import { Theme } from "@material-ui/core/styles/createMuiTheme";
import { transparentize } from "../utils";

const styles = (theme: Theme) =>
  createStyles({
    link: {
      textDecoration: "none"
    },
    background: {
      backgroundColor: transparentize(theme.palette.common.white, 0.5)
    }
  });

type MenuLink = {
  url: string;
  text: string;
};

type Props = {
  classes: Classes;
};

type State = {
  drawerOpen: boolean;
};

class Header extends React.Component<WithStyles<typeof styles>, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      drawerOpen: false
    };
    this.toggleMenu = this.toggleMenu.bind(this);
  }

  toggleMenu() {
    this.setState(prevState => ({
      drawerOpen: !prevState.drawerOpen
    }));
  }

  renderLink(link: MenuLink) {
    const { text, url } = link;
    return (
      <Link
        key={url}
        to={url}
        onClick={this.toggleMenu}
        className={this.props.classes.link}
      >
        <ListItem button>
          <ListItemText primary={text} />
        </ListItem>
      </Link>
    );
  }

  render() {
    const { drawerOpen } = this.state;
    const { classes } = this.props;

    const links = [
      { url: "/", text: "Home", icon: "home" },
      { url: "/history/", text: "History", icon: "history" },
      { url: "/exercises/", text: "Exercises", icon: "fitness" },
      { url: "/exercises/new/", text: "Create Exercise", icon: "create" }
    ];

    return (
      <div>
        <AppBar color="inherit" position="static">
          <Toolbar>
            <IconButton
              color="inherit"
              aria-label="Menu"
              onClick={this.toggleMenu}
            >
              <MenuIcon />
            </IconButton>
          </Toolbar>
        </AppBar>
        <Drawer
          anchor="left"
          open={drawerOpen}
          onClose={this.toggleMenu}
          ModalProps={{
            BackdropProps: {
              classes: {
                root: classes.background
              }
            }
          }}
        >
          <List>{links.map(link => this.renderLink(link))}</List>
        </Drawer>
      </div>
    );
  }
}

export default withStyles(styles)(Header);
