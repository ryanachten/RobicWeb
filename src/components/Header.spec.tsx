import React from "react";
import { createShallow } from "@material-ui/core/test-utils";
import Header from "./Header";
import renderer from "react-test-renderer";

describe("Component: Header", () => {
  let shallow: any;

  beforeEach(() => {
    shallow = createShallow({ untilSelector: "Header" });
  });

  it("renders correctly", () => {
    const tree = renderer.create(<Header />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should call toggleMenu", () => {
    const wrapper = shallow(<Header />);
    const instance = wrapper.instance();
    expect(instance.state.drawerOpen).toBeFalsy;
    instance.toggleMenu();
    expect(instance.state.drawerOpen).toBeTruthy;
    instance.toggleMenu();
    expect(instance.state.drawerOpen).toBeFalsy;
  });
});
