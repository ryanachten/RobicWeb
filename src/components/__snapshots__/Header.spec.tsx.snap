// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Header renders correctly 1`] = `
<div>
  <header
    className="MuiPaper-root-12 MuiPaper-elevation4-18 MuiAppBar-root-3 MuiAppBar-positionStatic-7"
  >
    <div
      className="MuiToolbar-root-39 MuiToolbar-regular-41 MuiToolbar-gutters-40"
    >
      <button
        aria-label="Menu"
        className="MuiButtonBase-root-49 MuiIconButton-root-43 MuiIconButton-colorInherit-44"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex="0"
        type="button"
      >
        <span
          className="MuiIconButton-label-48"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-52"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
            />
          </svg>
        </span>
        <span
          className="MuiTouchRipple-root-75"
        />
      </button>
    </div>
  </header>
</div>
`;
